File: C/86. Partition List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* partition(struct ListNode* head, int x) {
    struct ListNode smaller, greater;
    struct ListNode *cur, *scur, *gcur;
    
    cur = head;
    smaller.next = NULL;
    greater.next = NULL;
    scur = &smaller;
    gcur = &greater;

    while (cur != NULL) {
        if (cur->val < x) {
            scur->next = cur;
            scur = cur;
        } else {
            gcur->next = cur;
            gcur = cur;
        }
        cur = cur->next;
    }

    scur->next = greater.next;
    gcur->next = NULL;
    return smaller.next;
}

File: Python3/86. Partition List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        smaller, greater = ListNode(0), ListNode(0)
        stail, gtail = smaller, greater
        cur = head
        while cur:
            if cur.val < x:
                stail.next, stail = cur, cur
            else:
                gtail.next, gtail = cur, cur
            cur = cur.next
        stail.next, gtail.next = greater.next, None
        return smaller.next

